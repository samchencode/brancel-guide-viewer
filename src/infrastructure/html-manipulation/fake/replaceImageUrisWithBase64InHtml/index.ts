export { factory } from '@/infrastructure/html-manipulation/fake/replaceImageUrisWithBase64InHtml/replaceImageUrisWithBase64InHtml';
