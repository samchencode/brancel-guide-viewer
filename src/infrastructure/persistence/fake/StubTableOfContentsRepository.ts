import type { TableOfContentsRepository } from '@/domain/models/TableOfContents';
import {
  TableOfContents,
  TableOfContentsItem,
} from '@/domain/models/TableOfContents';

class StubTableOfContentsRepository implements TableOfContentsRepository {
  async get(): Promise<TableOfContents> {
    return new TableOfContents([
      new TableOfContentsItem(
        'Resurrection, A',
        'bbf46939-a2eb-493c-9b64-c2b8c00203b2'
      ),
      new TableOfContentsItem(
        'Radioactive Dreams',
        '724c9644-4bb1-4dc9-8131-edaeb0d0fc17'
      ),
      new TableOfContentsItem(
        'Death by China ',
        'df2dc7c1-7324-4418-ab41-100af5681ca8'
      ),
      new TableOfContentsItem(
        'Pont du Nord, Le',
        '313299af-d447-45a2-a3ed-63725ee5ca7c'
      ),
    ]);
  }
}

export { StubTableOfContentsRepository };
