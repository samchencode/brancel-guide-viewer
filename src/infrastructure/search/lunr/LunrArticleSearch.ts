import type {
  ArticleRepository,
  ArticleSearch,
  ArticleMatchData,
} from '@/domain/models/Article';
import { ArticleSearchResult } from '@/domain/models/Article/ArticleSearchResult';
import type { SearchableArticle } from '@/domain/models/Article/SearchableArticle';
import lunr from 'lunr';

type LunrMatchFieldMetadata = {
  position: [number, number][];
};

type LunrMatchMetadata = {
  [v: string]: {
    title: LunrMatchFieldMetadata;
    body: LunrMatchFieldMetadata;
  };
};

function metadataToMatchRanges(meta: LunrMatchMetadata): ArticleMatchData {
  const groups = Object.values(meta);
  return groups.reduce<ArticleMatchData>(
    (ag, v) => ({
      title: [...ag.title, ...v.title.position],
      body: [...ag.body, ...v.body.position],
    }),
    { title: [], body: [] }
  );
}

class LunrArticleSearch implements ArticleSearch {
  searchableArticles: Promise<SearchableArticle[]>;

  idx: Promise<lunr.Index>;

  constructor(private articleRepository: ArticleRepository) {
    this.searchableArticles = this.articleRepository.getAllSearchable();
    this.idx = this.buildIndex();
  }

  private async buildIndex() {
    const articles = await this.searchableArticles;

    return lunr(function config() {
      this.metadataWhitelist = ['position'];
      this.ref('id');
      this.field('title');
      this.field('body');

      for (const article of articles) {
        this.add(article);
      }
    });
  }

  async search(searchText: string): Promise<ArticleSearchResult[]> {
    const articles = await this.searchableArticles;
    const idx = await this.idx;
    const lunrResults = idx.search(searchText);
    return lunrResults.map((v) => {
      const article = articles.find((a) => a.id === v.ref);
      if (!article) throw new Error('Search result ref not found!');
      const metadata = v.matchData.metadata as LunrMatchMetadata;
      const matchRanges = metadataToMatchRanges(metadata);
      return new ArticleSearchResult(article, matchRanges, v.score);
    });
  }
}

export { LunrArticleSearch };
