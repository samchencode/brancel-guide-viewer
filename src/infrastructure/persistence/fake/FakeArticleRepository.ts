import { Article, ArticleId } from '@/domain/models/Article';
import type { ArticleRepository } from '@/domain/models/Article';
import type { FakeArticle } from '@/infrastructure/persistence/fake/fakeArticles';
import { fakeArticles } from '@/infrastructure/persistence/fake/fakeArticles';
import { RichText } from '@/domain/models/RichText/RichText';
import { sanitizeHtml } from '@/vendor/sanitizeHtml';

class FakeArticleRepository implements ArticleRepository {
  fakeArticles = fakeArticles.slice();

  async getAll(): Promise<Article[]> {
    return this.fakeArticles.map(this.makeArticle);
  }

  async getById(id: ArticleId): Promise<Article> {
    const article = this.fakeArticles.find((a) => a.id === id.toString());
    if (!article) throw Error(`Article with id of [${id}] not found`);
    return this.makeArticle(article);
  }

  private makeArticle(a: FakeArticle) {
    return new Article(
      new ArticleId(a.id),
      a.title,
      new RichText(sanitizeHtml, a.body)
    );
  }
}

export { FakeArticleRepository };
