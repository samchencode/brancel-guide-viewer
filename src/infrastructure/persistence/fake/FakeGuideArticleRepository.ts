import type {
  Article,
  ArticleId,
  ArticleRepository,
} from '@/domain/models/Article';
import type { SanitizeHtml } from '@/domain/models/RichText';
import { CheerioGuideParser } from '@/infrastructure/parsing/cheerio/CheerioGuideParser';
import type { GuideParser } from '@/domain/models/Guide';
import { stubGuide } from '@/infrastructure/persistence/fake/stubGuide';

class FakeGuideArticleRepository implements ArticleRepository {
  private articles: Article[];

  private guideParser: GuideParser;

  constructor(sanitizeHtml: SanitizeHtml) {
    this.guideParser = new CheerioGuideParser(sanitizeHtml);
    this.articles = this.guideParser.getArticles(stubGuide);
  }

  async getAll(): Promise<Article[]> {
    return this.articles;
  }

  async getById(id: ArticleId): Promise<Article> {
    const article = this.articles.find((a) => a.id.is(id));
    if (!article) throw Error(`article with id of ${id.toString()} not found!`);
    return article;
  }
}

export { FakeGuideArticleRepository };
