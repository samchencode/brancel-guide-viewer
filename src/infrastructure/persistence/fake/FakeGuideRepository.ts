import type { GuideParser, GuideRepository } from '@/domain/models/Guide';
import { Guide } from '@/domain/models/Guide';
import { stubGuide } from '@/infrastructure/persistence/fake/stubGuide';

class FakeGuideRepository implements GuideRepository {
  constructor(private readonly guideParser: GuideParser) {}

  async get() {
    return new Guide(stubGuide, this.guideParser);
  }

  async getLastUpdatedTimestamp(): Promise<Date> {
    return new Date(0);
  }
}

export { FakeGuideRepository };
