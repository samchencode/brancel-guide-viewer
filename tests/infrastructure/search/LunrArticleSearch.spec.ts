import type { ArticleRepository } from '@/domain/models/Article';
import { FakeArticleRepository } from '@/infrastructure/persistence/fake/FakeArticleRepository';
import { LunrArticleSearch } from '@/infrastructure/search/lunr/LunrArticleSearch';

describe('LunrArticleSearch', () => {
  describe('Instantiation', () => {
    it('should be created with article repo', () => {
      const repo = new FakeArticleRepository();
      const create = () => new LunrArticleSearch(repo);
      expect(create).not.toThrowError();
    });
  });

  describe('Behavior', () => {
    let repo: ArticleRepository;

    beforeEach(() => {
      repo = new FakeArticleRepository();
    });

    it('should get articles matching search term without html tags', async () => {
      const lunrSearch = new LunrArticleSearch(repo);

      const results = await lunrSearch.search('Hai My Article');
      const result = results[0];
      expect(result).toBeDefined();
      expect(result.article.body).not.toContain('<img');
      expect(results).toMatchSnapshot();
    });
  });
});
