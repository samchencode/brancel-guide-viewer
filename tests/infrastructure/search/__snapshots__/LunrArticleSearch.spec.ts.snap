// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LunrArticleSearch Behavior should get articles matching search term without html tags 1`] = `
[
  ArticleSearchResult {
    "article": {
      "body": "Hai my article Praesent blandit. Nam nulla. Integer pede justo, lacinia eget, tincidunt eget, tempus vel, pede. Morbi porttitor lorem id ligula. Suspendisse ornare consequat lectus. In est risus, auctor sed, tristique in, tempus sit amet, sem. Fusce consequat. Nulla nisl. Nunc nisl. Duis bibendum, felis sed interdum venenatis, turpis enim blandit mi, in porttitor pede justo eu massa. Donec dapibus. Duis at velit eu est congue elementum. In hac habitasse platea dictumst. Morbi vestibulum, velit id pretium iaculis, diam erat fermentum justo, nec condimentum neque sapien placerat ante. Nulla justo. Aliquam quis turpis eget elit sodales scelerisque. Mauris sit amet eros. Suspendisse accumsan tortor quis turpis. Sed ante. Vivamus tortor. Duis mattis egestas metus. Aenean fermentum. Donec ut mauris eget massa tempor convallis. Nulla neque libero, convallis eget, eleifend luctus, ultricies eu, nibh. Quisque id justo sit amet sapien dignissim vestibulum. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Nulla dapibus dolor vel est. Donec odio justo, sollicitudin ut, suscipit a, feugiat et, eros. Vestibulum ac est lacinia nisi venenatis tristique. Fusce congue, diam id ornare imperdiet, sapien urna pretium nisl, ut volutpat sapien arcu sed augue. Aliquam erat volutpat. In congue. Etiam justo. Etiam pretium iaculis justo. In hac habitasse platea dictumst. Etiam faucibus cursus urna. Ut tellus. Nulla ut erat id mauris vulputate elementum. Nullam varius. Nulla facilisi. Cras non velit nec nisi vulputate nonummy. Maecenas tincidunt lacus at velit. Vivamus vel nulla eget eros elementum pellentesque. Quisque porta volutpat erat. Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla. Nunc purus. Phasellus in felis. Donec semper sapien a libero. Nam dui. Proin leo odio, porttitor id, consequat in, consequat ut, nulla. Sed accumsan felis. Ut at dolor quis odio consequat varius. Integer ac leo. Pellentesque ultrices mattis odio. Donec vitae nisi.",
      "id": "bbf46939-a2eb-493c-9b64-c2b8c00203b2",
      "title": "Resurrection, A",
    },
    "matchData": {
      "body": [
        [
          0,
          3,
        ],
        [
          7,
          7,
        ],
      ],
      "title": [],
    },
    "score": 8.817621561396248,
  },
]
`;
